### PR TITLE
Histogram visualisation

### LINKED ISSUE #8145: [FR] Histogram plot logging over time (similar to Tensorboard and Weights and Biases)
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

I have found that MLFlow meets virtually every use case for model tracking and monitoring at my org except one - histogram logging over time.

Both Tensorboard and WandB (weights and biases) have the ability to generate a live readout of histogram plots as training progresses.

It turns out this is a very useful feature for certain machine learning researchers who leverage this feature to monitor weight, bias, and gradient distributions over time which is useful for developing novel model architectures.

Right now MLflow mainly logs metrics as scalars in a relational database, but I would argue that there is a reasonable use-case to log certain things such as event logs directly into artifact storage in a format amenable to live monitoring. Tensorboard has become weaker and weaker for this use case over time due to the lack of experiment organization.

Ideally, MLFlow would be able to read and plot Tensorboard event files deposited in artifact storage directly. Perhaps `smart_log_metric` could intelligently log np.ndarray scalars as in artifact storage and scalars in the relational database, and the dashboard could conceivable plot both types of data cleanly.

### Motivation

> #### What is the use case for this feature?

> #### Why is this use case valuable to support for MLflow users in general?

> #### Why is this use case valuable to support for your project(s) or organization?

> #### Why is it currently difficult to achieve this use case?

### Details

_No response_

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [X] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by joelrobin18**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/joelrobin18/mlflow/pull/19093?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19093/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19093/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/19093/merge
```

</p>
</details>

### What changes are proposed in this pull request?

Histogram visualisation in mlflow. 

This PR introduces histogram logging and 3D visualization capabilities to MLflow, enabling users to track and visualize the evolution of weight and gradient distributions during model training. 
Backend: Implemented a new mlflow.log_histogram() fluent API that accepts either raw values (e.g., model weights/gradients) or pre-computed histogram data.
Frontend: A new 3D Histogram chart type renders distributions over training steps as interactive Plotly.js surface plots, accessible via the chart menu in the MLflow UI.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

image
Screenshot 2025-11-27 at 10 40 30 PM
Screenshot 2025-11-27 at 10 40 45 PM
Screenshot 2025-11-27 at 10 40 58 PM

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

MLflow now supports 3D visualisation on histogram out of the box.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/server/js/src/experiment-tracking/components/runs-charts/components/RunsChartsAddChartMenu.tsx [hunk 1]
```diff
@@ -8,6 +8,7 @@ import { ReactComponent as ChartParallelIcon } from '../../../../common/static/c
 import { ReactComponent as ChartScatterIcon } from '../../../../common/static/chart-scatter.svg';
 import { ReactComponent as ChartDifferenceIcon } from '../../../../common/static/chart-difference.svg';
 import { ReactComponent as ChartImageIcon } from '../../../../common/static/chart-image.svg';
+import { ReactComponent as ChartHistogramIcon } from '../../../../common/static/chart-histogram.svg';
```
Comment (daniellok-db) [inline mlflow/server/js/src/experiment-tracking/components/runs-charts/components/RunsChartsAddChartMenu.tsx]: there's a number of chart-related icons from design system, maybe <BarStackedIcon /> works? i think as far as possible it's nice to not rely on custom svgs

**Comment by rsamf** [comment]
@joelrobin18 @daniellok-db It seems like development of this has stopped since there hasn't been any activity for 2 months. Can we resume progress on this? It's important for me, and it would ultimately allow me switch over to mlflow for nearly 100% of my work.

File: mlflow/tracking/fluent.py [hunk 2]
```diff
@@ -1859,6 +1863,121 @@ def log_image(
     MlflowClient().log_image(run_id, image, artifact_file, key, step, timestamp, synchronous)
 
 
+def log_histogram(
+    values: Union["numpy.ndarray", list] | None = None,
+    *,
+    bins: list[float] | None = None,
+    counts: list[float] | None = None,
+    key: str,
```
Comment (B-Step62) [inline mlflow/tracking/fluent.py:1869]: Having required valeu after optional ones is confusing, can we move it up before "values"?

File: mlflow/tracking/fluent.py [hunk 3]
```diff
@@ -1859,6 +1863,121 @@ def log_image(
...
+    step: int | None = None,
+    timestamp: int | None = None,
+    num_bins: int = 30,
+) -> None:
+    """
+    Logs a histogram in MLflow for tracking distributions over time (e.g., weights,
+    gradients, activations during model training).
+
+    You can either:
+    1. Pass raw values and let MLflow compute the histogram
```
Comment (B-Step62) [inline mlflow/tracking/fluent.py]: Do we need to support both? It is a bit confusing since "bins" can be also interpreted as a parameter to specify number of buckets to categorize raw values (num_bins). Based on theoriginal feature request, we only need to handle raw data.

File: mlflow/utils/histogram_utils.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            bin_edges=data["bin_edges"],
+            counts=data["counts"],
+            min_value=data.get("min_value"),
+            max_value=data.get("max_value"),
+        )
+
+
+def compute_histogram_from_values(
+    values: np.ndarray | list,
+    num_bins: int = 30,
```
Comment (B-Step62) [inline mlflow/utils/histogram_utils.py]: nit: do we need default value here?

File: mlflow/utils/histogram_utils.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    Compute histogram from raw values.
+
+    Args:
+        values: Raw data values
+        num_bins: Number of bins to use
+
+    Returns:
+        Tuple of (bin_edges, counts, min_value, max_value)
+    """
+    values = np.asarray(values, dtype=np.float64)
```
Comment (B-Step62) [inline mlflow/utils/histogram_utils.py]: nit: Can we have this casting upstream and simplify this func only accept numpy array?

File: mlflow/utils/histogram_utils.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    Args:
+        histogram: HistogramData instance
+        file_path: Path to JSON file
+    """
+    file_path = Path(file_path)
+
+    # Load existing histograms if file exists
+    if file_path.exists():
+        try:
+            histograms = load_histograms_from_json(file_path)
```
Comment (B-Step62) [inline mlflow/utils/histogram_utils.py]: Similarly to the above comment, load-and-update can cause several troubles.

File: mlflow/utils/histogram_utils.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        bin_edges = np.array([min_val - 0.5, max_val + 0.5])
+        counts = np.array([float(len(valid_values))])
+    else:
+        counts, bin_edges = np.histogram(valid_values, bins=num_bins)
+        counts = counts.astype(np.float64)
+
+    return bin_edges, counts, min_val, max_val
+
+
+def save_histogram_to_json(histogram: HistogramData, file_path: str | Path) -> None:
```
Comment (B-Step62) [inline mlflow/utils/histogram_utils.py:111]: Looks like dead code

File: mlflow/tracking/fluent.py [hunk 8]
```diff
@@ -1859,6 +1863,121 @@ def log_image(
...
+            "Either pass raw 'values' to compute histogram, or pass pre-computed 'bins' and"
+            "'counts'."
+        )
+
+    if values is None and (bins is None or counts is None):
+        raise ValueError(
+            "Must provide either 'values' to compute histogram, "
+            "or both 'bins' and 'counts' for pre-computed histogram."
+        )
+
```
Comment (B-Step62) [inline mlflow/tracking/fluent.py]: Can we validate the type of values?

File: mlflow/tracking/client.py [hunk 9]
```diff
@@ -3254,6 +3254,110 @@ def log_image(
...
+        # Append histogram to JSON file using temporary file
+        with self._log_artifact_helper(run_id, artifact_file) as tmp_path:
+            # Check if file already exists and download it to append
+            artifact_dir = posixpath.dirname(artifact_file)
+            artifact_dir = None if artifact_dir == "" else artifact_dir
+            artifacts = [f.path for f in self.list_artifacts(run_id, path=artifact_dir)]
+
+            if artifact_file in artifacts:
+                with tempfile.TemporaryDirectory() as tmpdir:
+                    downloaded_path = self.download_artifacts(
```
Comment (B-Step62) [inline mlflow/tracking/client.py]: The load-and-save approach has several complexities need consideration:
- When the file becomes large, the oveahead of loading the existing file becomes large.
- When multiple nodes edit the histogram at the same time, it becomes broken state due to race condition.

Ideally we could use append write to the file and prevent race condition with a lock, but supporting various artifact path we support might be challenging. According to Tensorboard code, they use native append write for some storage (local, Azure, GCP), while using a different ad-hoc logic for S3 (buffer updates).

File: mlflow/server/js/src/experiment-tracking/components/runs-charts/components/RunsHistogram3DPlot.tsx [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+ * Converts histogram data into a 3D surface plot format
+ */
+const convertHistogramsTo3DSurface = (histograms: HistogramData[], logScale: boolean = false) => {
+  if (histograms.length === 0) {
+    return { x: [], y: [], z: [] };
+  }
+
+  const sortedHistograms = [...histograms].sort((a, b) => a.step - b.step);
+
+  const steps = sortedHistograms.map((h) => h.step);
```
Comment (B-Step62) [inline mlflow/server/js/src/experiment-tracking/components/runs-charts/components/RunsHistogram3DPlot.tsx]: What is the expected behavior when there is duplicated histograms for a single step?

File: mlflow/utils/histogram_utils.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        try:
+            histograms = load_histograms_from_json(file_path)
+        except Exception as e:
+            _logger.warning(f"Failed to load existing histograms from {file_path}: {e}")
+            histograms = []
+    else:
+        histograms = []
+
+    # Append new histogram
+    histograms.append(histogram)
```
Comment (B-Step62) [inline mlflow/utils/histogram_utils.py]: This requires reading->deserializing->serializing->writing the full json per every append, which is inefficient. Can we explore other format that is more efficient for appending like JSONL?

File: mlflow/tracking/fluent.py (hunk 3 above)
Comment (joelrobin18) [reply mlflow/tracking/fluent.py]: Removed bins/counts support entirely. log_histogram now only accepts raw values and computes the histogram internally.

File: mlflow/utils/histogram_utils.py (hunk 4 above)
Comment (joelrobin18) [reply mlflow/utils/histogram_utils.py]: Removed the default. compute_histogram_from_values now requires num_bins explicitly; the default of 30 lives only in the public log_histogram API.

File: mlflow/utils/histogram_utils.py (hunk 5 above)
Comment (joelrobin18) [reply mlflow/utils/histogram_utils.py]: Moved the np.asarray casting to fluent.py. compute_histogram_from_values now only accepts np.ndarray.

File: mlflow/utils/histogram_utils.py (hunk 7 above)
Comment (joelrobin18) [reply mlflow/utils/histogram_utils.py:111]: Removed them

File: mlflow/server/js/src/experiment-tracking/components/runs-charts/components/RunsHistogram3DPlot.tsx (hunk 10 above)
Comment (joelrobin18) [reply mlflow/server/js/src/experiment-tracking/components/runs-charts/components/RunsHistogram3DPlot.tsx]: Added deduplication — when multiple histograms exist for the same step, the last entry is kept. With the new per-step file storage, duplicates at the storage level are avoided since re-logging a step overwrites the file, but the frontend handles it defensively as well.

**Comment by italoaa** [comment]
This looks amazing, I am not familiar with the codebase at all but would love to help push this forward. @joelrobin18 do you have any place I could take a look?
